### DISCUSSION TITLE
Return a non-zero exit code when Renovate was unable to find a preset

### DISCUSSION THREAD

**Discussion by tenstan** [renovatebot/renovate#24891] [Suggest an Idea]
### What would you like Renovate to be able to do?

Renovate returns exit code 0 even when it could not find presets defined in `renovate.json` such as reported below:
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "path/to//my/preset"
  ]
}
```
```
INFO: Repository started (repository=mynamespace/myrepository)
       "renovateVersion": "34.27.1"
INFO: Throwing preset error (repository=mynamespace/myrepository)
       "validationError": "Cannot find preset's package (path/to//my/preset)"
INFO: Repository has invalid config (repository=mynamespace/myrepository)
       "error": {
         "validationError": "Cannot find preset's package (path/to//my/preset)",
         "message": "config-validation",
         "stack": "Error: config-validation\n    at fetchPreset (C:\\Data\\Agent\\_work\\139\\s\\node_modules\\renovate\\dist\\config\\presets\\index.js:293:23)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async Object.resolveConfigPresets (C:\\Data\\Agent\\_work\\139\\s\\node_modules\\renovate\\dist\\config\\presets\\index.js:238:39)\n    at async mergeRenovateConfig (C:\\Data\\Agent\\_work\\139\\s\\node_modules\\renovate\\dist\\workers\\repository\\init\\merge.js:207:61)\n    at async getRepoConfig (C:\\Data\\Agent\\_work\\139\\s\\node_modules\\renovate\\dist\\workers\\repository\\init\\config.js:15:14)\n    at async initRepo (C:\\Data\\Agent\\_work\\139\\s\\node_modules\\renovate\\dist\\workers\\repository\\init\\index.js:33:14)\n    at async Object.renovateRepository (C:\\Data\\Agent\\_work\\139\\s\\node_modules\\renovate\\dist\\workers\\repository\\index.js:43:18)\n    at async attributes.repository (C:\\Data\\Agent\\_work\\139\\s\\node_modules\\renovate\\dist\\workers\\global\\index.js:132:17)\n    at async Object.start (C:\\Data\\Agent\\_work\\139\\s\\node_modules\\renovate\\dist\\workers\\global\\index.js:121:13)\n    at async C:\\Data\\Agent\\_work\\139\\s\\node_modules\\renovate\\dist\\renovate.js:16:24"
       }
```

During CI/CD these validation errors can be hard to spot. The pipeline reports it succeeded due to the exit code. Preset locations can change outside of your control and you will have to go through the logs to see what went wrong.

My suggestion is to return a non-zero exit code when Renovate was unable to locate a preset.

### If you have any ideas on how this should be implemented, please tell us here.

I believe Renovate only returns a non-zero exit code when an error is logged. Since the output shows an error occurred during validation which interestingly gets posted as an INFO log, perhaps this information can be changed to be logged as an error? I'm unsure whether Renovate continues to create PRs even when it was unable to find presets, so this might unnecessarily break the process.

If this is too breaking, perhaps a configuration option could enable this functionality either on `renovate` or `renovate-config-validator`.

### Is this a feature you are interested in implementing yourself?

Yes

**Comment by rarkins** [comment]
A missing preset is considered part of a wider problem of config validation error. Would you like a non-zero exit code for any case of config validation error?

BTW I'm surprised if there's no WARN thrown. I recommend you monitor both errors and warnings in your logs.

**Comment by tenstan** [comment]
I'd personally prefer if all config validation errors would return a non-zero exit code, but I'm not sure with the wide capabilities of Renovate if that is appropriate. In the case of missing presets for example, configuration can still be valid even if there was a missing preset, so I was unsure if this was by design anyway. Like I said though, errors can then be hard to detect when you are automating Renovate in CI/CD and the sorts.

About the missing WARNs, Renovate does not seem to log any warnings/errors when it detects a preset error. [I've got a working sample here in Azure Pipelines.](https://dev.azure.com/smarten123/renovate-poc/_build/results?buildId=68&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=efa2bfe1-554a-50c8-79b6-ef106ad3c7c2) As you can see, the warning about `ensureIssue()` does get logged.

**Comment by rarkins** [comment]
Here's what I think the possible steps are:
1. Ensure we have a meaningful WARN log by default for configuration validation errors, assuming it's correct that we don't have one today
2. Add a new global config option `configValidationError` or similar which allows users such as @ovalice to opt into logging and error in such cases and therefore exiting with non-zero

**Comment by boneskull** [comment]
I was bit by this.  What I'd really like to see is something in the `renovate` CLI which just validates a config file and returns nonzero if it fails.  This is especially important for shared configs.

**Comment by HonkingGoose** [comment]
> Here's what I think the possible steps are:
> 
>     1. Ensure we have a meaningful WARN log by default for configuration validation errors, assuming it's correct that we don't have one today
> 
>     2. Add a new global config option `configValidationError` or similar which allows users such as @ovalice to opt into logging and error in such cases and therefore exiting with non-zero

Are we `status:ready` then, or do we need to think about this more?

**Reply by reduckted** [reply]
is this something that you'll accept a PR for?

**Reply by reduckted** [reply]
I've created https://github.com/renovatebot/renovate/pull/39177 for the `configValidationError` option. 

The warnings for configuration validation errors was conveniently changed just last week. 🎉 
- #38924
- #38925

**Reply by RahulGautamSingh** [reply]
Will take a look. Thanks!
